src/core/jit/graphql_executor.rs

```diff
@@ -45,25 +74,25 @@ impl Executor for JITExecutor {
     fn execute(&self, request: async_graphql::Request) -> impl Future<Output = Response> + Send {
         let jit_request = jit::Request::from(request);
 
-        async {
-            match ConstValueExecutor::new(&jit_request, self.app_ctx.clone()) {
+        async move {
+            match ConstValueExecutor::new(&jit_request, &self.app_ctx) {
                 Ok(exec) => {
-                    let is_introspection_query =
-                        self.app_ctx.blueprint.server.get_enable_introspection()
-                            && exec.plan.is_introspection_query;
-
-                    let jit_resp = exec
-                        .execute(&self.req_ctx, &jit_request)
-                        .await
-                        .into_async_graphql();
-
-                    if is_introspection_query {
-                        let async_req =
-                            async_graphql::Request::from(jit_request).only_introspection();
-                        let async_resp = self.app_ctx.execute(async_req).await;
-                        jit_resp.merge_right(async_resp)
+                    if self.is_query && exec.plan.dedupe {
+                        let out = self
+                            .app_ctx
+                            .dedupe_operation_handler
+                            .dedupe(&self.operation_id, || {
+                                Box::pin(async move {
+                                    let resp = self.exec(exec, jit_request).await;
+                                    Ok(Arc::new(resp))
+                                })
+                            })
+                            .await;
+                        // TODO: check if `unwrap_or_default` is correct
+                        let val = out.unwrap_or_default();
+                        Arc::into_inner(val).unwrap_or_default()
```

Comment:
Use `Response::from_errors` or something equivalent to return incase into_inner doesn't work.